src/snowflake/connector/config_parser/parser.py

```diff
@@ -0,0 +1,188 @@
+            e = e[k]
+        if isinstance(e, (Table, TOMLDocument)):
+            # If we got a TOML table we probably want it in dictionary form
+            return e.value
+        return e
+
+
+class ConfigParser:
+    def __init__(
+        self,
+        *,
+        name: str,
+        file_path: Path | None = None,
+    ):
+        self.name = name
+        self.file_path = file_path
+        # Objects holding subparsers and options
+        self._options: dict[str, ConfigOption] = dict()
+        self._sub_parsers: dict[str, ConfigParser] = dict()
+        # Dictionary to cache read in config file
+        self._conf: TOMLDocument | None = None
+        # Information necessary to be able to nest elements
+        #  and add options in O(1)
+        self._root_parser: ConfigParser = self
+        self._nest_path = [name]
+
+    def read_config(
+        self,
+    ) -> None:
+        """Read and parse config file."""
+        if self.file_path is None:
+            raise ConfigParserError(
+                "ConfigParser is trying to read config file, but it doesn't have one"
+            )
+        LOGGER.debug(f"reading configuration file from {str(self.file_path)}")
+        try:
+            self._conf = tomlkit.parse(self.file_path.read_text())
+        except Exception as e:
+            raise ConfigSourceError(
+                f"An unknown error happened while loading '{str(self.file_path)}"
+                f"', please see the error: {e}"
+            )
+
+    @wraps(ConfigOption.__init__)
+    def add_option(
+        self,
+        *args,
+        **kwargs,
+    ) -> None:
+        kwargs["_root_parser"] = self._root_parser
+        kwargs["_nest_path"] = self._nest_path
+        new_option = ConfigOption(
+            *args,
+            **kwargs,
+        )
+        self._check_child_conflict(new_option.name)
+        self._options[new_option.name] = new_option
+
+    def _check_child_conflict(self, name: str) -> None:
+        if name in (self._options.keys() | self._sub_parsers.keys()):
+            raise ConfigParserError(
+                f"'{name}' subparser, or option conflicts with a child element of '{self.name}'"
+            )
+
+    def add_subparser(self, other: ConfigParser) -> None:
+        self._check_child_conflict(other.name)
+        self._sub_parsers[other.name] = other
+
+        def _root_setter_helper(node: ConfigParser):
+            # Deal with ConfigParsers
+            node._root_parser = self._root_parser
+            node._nest_path = self._nest_path + node._nest_path
+            for sub_parser in node._sub_parsers.values():
+                _root_setter_helper(sub_parser)
+            # Deal with ConfigOptions
+            for option in node._options.values():
+                option._root_parser = self._root_parser
+                option._nest_path = self._nest_path + option._nest_path
+
+        _root_setter_helper(other)
+
+    def __getitem__(self, item: str) -> ConfigOption | ConfigParser:
```

Comment:
since we have `__getitem__`, shall we have the counter part `__setitem__`?

Reply:
For now `ConfigParser` is only for reading values. In the future we can add this, but just not for now